{
 "cells": [
  {
   "cell_type": "markdown",
   "id": "b1710dae-970a-4acf-b2db-e5fbdcff3f14",
   "metadata": {},
   "source": [
    "# Oanda Error Debuging"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "id": "c7e60fa8-b23c-41b7-8695-f5e64aabae6c",
   "metadata": {},
   "outputs": [],
   "source": [
    "import pandas as pd\n",
    "import numpy as np\n",
    "import tpqoa\n",
    "from datetime import datetime, timedelta, timezone # Need to import timezone seperatly\n",
    "import time\n",
    "import v20.errors\n",
    "import traceback"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "id": "41de7f2f-fe39-459a-b69c-a044bf94a716",
   "metadata": {},
   "outputs": [],
   "source": [
    "import warnings\n",
    "warnings.simplefilter(action='ignore', category=FutureWarning)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "id": "4a997c15-e5ac-4f13-a49d-734a6a91668e",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "\n",
      "--- Entering get_most_recent ---\n",
      "\n",
      "Loop Iteration: 1\n",
      "Fetching history from 2025-04-07 02:46:53 UTC to 2025-04-12 02:46:53 UTC\n",
      "ERROR during get_history: 400 response for GET https://api-fxpractice.oanda.com:443/v3/instruments/EUR_USD/candles?price=M&granularity=S5&from=2025-04-07T02%3A46%3A53.527907%2B00%3A00.000000000Z&to=2025-04-07T08%3A46%3A53.527907%2B00%3A00.000000000Z does not have field 'candles' (contains 'errorMessage')\n",
      "      This often indicates invalid request parameters (e.g., dates).\n",
      "      *** PLEASE VERIFY YOUR SYSTEM CLOCK IS CORRECT! ***\n",
      "Waiting before retry...\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Traceback (most recent call last):\n",
      "  File \"C:\\Users\\audri\\AppData\\Local\\Temp\\ipykernel_26804\\3262540947.py\", line 53, in get_most_recent\n",
      "    df = self.get_history(instrument = self.instrument, start = past_aware, end = now_aware,\n",
      "         ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^\n",
      "  File \"C:\\Users\\audri\\anaconda3\\Lib\\site-packages\\tpqoa\\tpqoa.py\", line 232, in get_history\n",
      "    batch = self.retrieve_data(instrument, batch_start, batch_end,\n",
      "            ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^\n",
      "  File \"C:\\Users\\audri\\anaconda3\\Lib\\site-packages\\tpqoa\\tpqoa.py\", line 167, in retrieve_data\n",
      "    raw = raw.get('candles')\n",
      "          ^^^^^^^^^^^^^^^^^^\n",
      "  File \"C:\\Users\\audri\\anaconda3\\Lib\\site-packages\\v20\\response.py\", line 35, in get\n",
      "    raise ResponseNoField(self, field)\n",
      "v20.errors.ResponseNoField: 400 response for GET https://api-fxpractice.oanda.com:443/v3/instruments/EUR_USD/candles?price=M&granularity=S5&from=2025-04-07T02%3A46%3A53.527907%2B00%3A00.000000000Z&to=2025-04-07T08%3A46%3A53.527907%2B00%3A00.000000000Z does not have field 'candles' (contains 'errorMessage')\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "\n",
      "Loop Iteration: 2\n",
      "Fetching history from 2025-04-07 02:47:05 UTC to 2025-04-12 02:47:05 UTC\n",
      "ERROR during get_history: 400 response for GET https://api-fxpractice.oanda.com:443/v3/instruments/EUR_USD/candles?price=M&granularity=S5&from=2025-04-07T02%3A47%3A05.979775%2B00%3A00.000000000Z&to=2025-04-07T08%3A47%3A05.979775%2B00%3A00.000000000Z does not have field 'candles' (contains 'errorMessage')\n",
      "      This often indicates invalid request parameters (e.g., dates).\n",
      "      *** PLEASE VERIFY YOUR SYSTEM CLOCK IS CORRECT! ***\n",
      "Waiting before retry...\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Traceback (most recent call last):\n",
      "  File \"C:\\Users\\audri\\AppData\\Local\\Temp\\ipykernel_26804\\3262540947.py\", line 53, in get_most_recent\n",
      "    df = self.get_history(instrument = self.instrument, start = past_aware, end = now_aware,\n",
      "         ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^\n",
      "  File \"C:\\Users\\audri\\anaconda3\\Lib\\site-packages\\tpqoa\\tpqoa.py\", line 232, in get_history\n",
      "    batch = self.retrieve_data(instrument, batch_start, batch_end,\n",
      "            ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^\n",
      "  File \"C:\\Users\\audri\\anaconda3\\Lib\\site-packages\\tpqoa\\tpqoa.py\", line 167, in retrieve_data\n",
      "    raw = raw.get('candles')\n",
      "          ^^^^^^^^^^^^^^^^^^\n",
      "  File \"C:\\Users\\audri\\anaconda3\\Lib\\site-packages\\v20\\response.py\", line 35, in get\n",
      "    raise ResponseNoField(self, field)\n",
      "v20.errors.ResponseNoField: 400 response for GET https://api-fxpractice.oanda.com:443/v3/instruments/EUR_USD/candles?price=M&granularity=S5&from=2025-04-07T02%3A47%3A05.979775%2B00%3A00.000000000Z&to=2025-04-07T08%3A47%3A05.979775%2B00%3A00.000000000Z does not have field 'candles' (contains 'errorMessage')\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "\n",
      "Loop Iteration: 3\n",
      "Fetching history from 2025-04-07 02:47:18 UTC to 2025-04-12 02:47:18 UTC\n",
      "ERROR during get_history: 400 response for GET https://api-fxpractice.oanda.com:443/v3/instruments/EUR_USD/candles?price=M&granularity=S5&from=2025-04-07T02%3A47%3A18.110718%2B00%3A00.000000000Z&to=2025-04-07T08%3A47%3A18.110718%2B00%3A00.000000000Z does not have field 'candles' (contains 'errorMessage')\n",
      "      This often indicates invalid request parameters (e.g., dates).\n",
      "      *** PLEASE VERIFY YOUR SYSTEM CLOCK IS CORRECT! ***\n",
      "Waiting before retry...\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Traceback (most recent call last):\n",
      "  File \"C:\\Users\\audri\\AppData\\Local\\Temp\\ipykernel_26804\\3262540947.py\", line 53, in get_most_recent\n",
      "    df = self.get_history(instrument = self.instrument, start = past_aware, end = now_aware,\n",
      "         ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^\n",
      "  File \"C:\\Users\\audri\\anaconda3\\Lib\\site-packages\\tpqoa\\tpqoa.py\", line 232, in get_history\n",
      "    batch = self.retrieve_data(instrument, batch_start, batch_end,\n",
      "            ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^\n",
      "  File \"C:\\Users\\audri\\anaconda3\\Lib\\site-packages\\tpqoa\\tpqoa.py\", line 167, in retrieve_data\n",
      "    raw = raw.get('candles')\n",
      "          ^^^^^^^^^^^^^^^^^^\n",
      "  File \"C:\\Users\\audri\\anaconda3\\Lib\\site-packages\\v20\\response.py\", line 35, in get\n",
      "    raise ResponseNoField(self, field)\n",
      "v20.errors.ResponseNoField: 400 response for GET https://api-fxpractice.oanda.com:443/v3/instruments/EUR_USD/candles?price=M&granularity=S5&from=2025-04-07T02%3A47%3A18.110718%2B00%3A00.000000000Z&to=2025-04-07T08%3A47%3A18.110718%2B00%3A00.000000000Z does not have field 'candles' (contains 'errorMessage')\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "\n",
      "Loop Iteration: 4\n",
      "Fetching history from 2025-04-07 02:47:30 UTC to 2025-04-12 02:47:30 UTC\n",
      "ERROR during get_history: 400 response for GET https://api-fxpractice.oanda.com:443/v3/instruments/EUR_USD/candles?price=M&granularity=S5&from=2025-04-07T02%3A47%3A30.244922%2B00%3A00.000000000Z&to=2025-04-07T08%3A47%3A30.244922%2B00%3A00.000000000Z does not have field 'candles' (contains 'errorMessage')\n",
      "      This often indicates invalid request parameters (e.g., dates).\n",
      "      *** PLEASE VERIFY YOUR SYSTEM CLOCK IS CORRECT! ***\n",
      "Waiting before retry...\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Traceback (most recent call last):\n",
      "  File \"C:\\Users\\audri\\AppData\\Local\\Temp\\ipykernel_26804\\3262540947.py\", line 53, in get_most_recent\n",
      "    df = self.get_history(instrument = self.instrument, start = past_aware, end = now_aware,\n",
      "         ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^\n",
      "  File \"C:\\Users\\audri\\anaconda3\\Lib\\site-packages\\tpqoa\\tpqoa.py\", line 232, in get_history\n",
      "    batch = self.retrieve_data(instrument, batch_start, batch_end,\n",
      "            ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^\n",
      "  File \"C:\\Users\\audri\\anaconda3\\Lib\\site-packages\\tpqoa\\tpqoa.py\", line 167, in retrieve_data\n",
      "    raw = raw.get('candles')\n",
      "          ^^^^^^^^^^^^^^^^^^\n",
      "  File \"C:\\Users\\audri\\anaconda3\\Lib\\site-packages\\v20\\response.py\", line 35, in get\n",
      "    raise ResponseNoField(self, field)\n",
      "v20.errors.ResponseNoField: 400 response for GET https://api-fxpractice.oanda.com:443/v3/instruments/EUR_USD/candles?price=M&granularity=S5&from=2025-04-07T02%3A47%3A30.244922%2B00%3A00.000000000Z&to=2025-04-07T08%3A47%3A30.244922%2B00%3A00.000000000Z does not have field 'candles' (contains 'errorMessage')\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "\n",
      "Loop Iteration: 5\n",
      "Fetching history from 2025-04-07 02:47:42 UTC to 2025-04-12 02:47:42 UTC\n",
      "ERROR during get_history: 400 response for GET https://api-fxpractice.oanda.com:443/v3/instruments/EUR_USD/candles?price=M&granularity=S5&from=2025-04-07T02%3A47%3A42.458901%2B00%3A00.000000000Z&to=2025-04-07T08%3A47%3A42.458901%2B00%3A00.000000000Z does not have field 'candles' (contains 'errorMessage')\n",
      "      This often indicates invalid request parameters (e.g., dates).\n",
      "      *** PLEASE VERIFY YOUR SYSTEM CLOCK IS CORRECT! ***\n",
      "Waiting before retry...\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Traceback (most recent call last):\n",
      "  File \"C:\\Users\\audri\\AppData\\Local\\Temp\\ipykernel_26804\\3262540947.py\", line 53, in get_most_recent\n",
      "    df = self.get_history(instrument = self.instrument, start = past_aware, end = now_aware,\n",
      "         ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^\n",
      "  File \"C:\\Users\\audri\\anaconda3\\Lib\\site-packages\\tpqoa\\tpqoa.py\", line 232, in get_history\n",
      "    batch = self.retrieve_data(instrument, batch_start, batch_end,\n",
      "            ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^\n",
      "  File \"C:\\Users\\audri\\anaconda3\\Lib\\site-packages\\tpqoa\\tpqoa.py\", line 167, in retrieve_data\n",
      "    raw = raw.get('candles')\n",
      "          ^^^^^^^^^^^^^^^^^^\n",
      "  File \"C:\\Users\\audri\\anaconda3\\Lib\\site-packages\\v20\\response.py\", line 35, in get\n",
      "    raise ResponseNoField(self, field)\n",
      "v20.errors.ResponseNoField: 400 response for GET https://api-fxpractice.oanda.com:443/v3/instruments/EUR_USD/candles?price=M&granularity=S5&from=2025-04-07T02%3A47%3A42.458901%2B00%3A00.000000000Z&to=2025-04-07T08%3A47%3A42.458901%2B00%3A00.000000000Z does not have field 'candles' (contains 'errorMessage')\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "\n",
      "Loop Iteration: 6\n",
      "Fetching history from 2025-04-07 02:47:54 UTC to 2025-04-12 02:47:54 UTC\n",
      "ERROR during get_history: 400 response for GET https://api-fxpractice.oanda.com:443/v3/instruments/EUR_USD/candles?price=M&granularity=S5&from=2025-04-07T02%3A47%3A54.597610%2B00%3A00.000000000Z&to=2025-04-07T08%3A47%3A54.597610%2B00%3A00.000000000Z does not have field 'candles' (contains 'errorMessage')\n",
      "      This often indicates invalid request parameters (e.g., dates).\n",
      "      *** PLEASE VERIFY YOUR SYSTEM CLOCK IS CORRECT! ***\n",
      "Waiting before retry...\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Traceback (most recent call last):\n",
      "  File \"C:\\Users\\audri\\AppData\\Local\\Temp\\ipykernel_26804\\3262540947.py\", line 53, in get_most_recent\n",
      "    df = self.get_history(instrument = self.instrument, start = past_aware, end = now_aware,\n",
      "         ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^\n",
      "  File \"C:\\Users\\audri\\anaconda3\\Lib\\site-packages\\tpqoa\\tpqoa.py\", line 232, in get_history\n",
      "    batch = self.retrieve_data(instrument, batch_start, batch_end,\n",
      "            ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^\n",
      "  File \"C:\\Users\\audri\\anaconda3\\Lib\\site-packages\\tpqoa\\tpqoa.py\", line 167, in retrieve_data\n",
      "    raw = raw.get('candles')\n",
      "          ^^^^^^^^^^^^^^^^^^\n",
      "  File \"C:\\Users\\audri\\anaconda3\\Lib\\site-packages\\v20\\response.py\", line 35, in get\n",
      "    raise ResponseNoField(self, field)\n",
      "v20.errors.ResponseNoField: 400 response for GET https://api-fxpractice.oanda.com:443/v3/instruments/EUR_USD/candles?price=M&granularity=S5&from=2025-04-07T02%3A47%3A54.597610%2B00%3A00.000000000Z&to=2025-04-07T08%3A47%3A54.597610%2B00%3A00.000000000Z does not have field 'candles' (contains 'errorMessage')\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "\n",
      "Loop Iteration: 7\n",
      "Fetching history from 2025-04-07 02:48:06 UTC to 2025-04-12 02:48:06 UTC\n",
      "ERROR during get_history: 400 response for GET https://api-fxpractice.oanda.com:443/v3/instruments/EUR_USD/candles?price=M&granularity=S5&from=2025-04-07T02%3A48%3A06.747334%2B00%3A00.000000000Z&to=2025-04-07T08%3A48%3A06.747334%2B00%3A00.000000000Z does not have field 'candles' (contains 'errorMessage')\n",
      "      This often indicates invalid request parameters (e.g., dates).\n",
      "      *** PLEASE VERIFY YOUR SYSTEM CLOCK IS CORRECT! ***\n",
      "Waiting before retry...\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Traceback (most recent call last):\n",
      "  File \"C:\\Users\\audri\\AppData\\Local\\Temp\\ipykernel_26804\\3262540947.py\", line 53, in get_most_recent\n",
      "    df = self.get_history(instrument = self.instrument, start = past_aware, end = now_aware,\n",
      "         ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^\n",
      "  File \"C:\\Users\\audri\\anaconda3\\Lib\\site-packages\\tpqoa\\tpqoa.py\", line 232, in get_history\n",
      "    batch = self.retrieve_data(instrument, batch_start, batch_end,\n",
      "            ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^\n",
      "  File \"C:\\Users\\audri\\anaconda3\\Lib\\site-packages\\tpqoa\\tpqoa.py\", line 167, in retrieve_data\n",
      "    raw = raw.get('candles')\n",
      "          ^^^^^^^^^^^^^^^^^^\n",
      "  File \"C:\\Users\\audri\\anaconda3\\Lib\\site-packages\\v20\\response.py\", line 35, in get\n",
      "    raise ResponseNoField(self, field)\n",
      "v20.errors.ResponseNoField: 400 response for GET https://api-fxpractice.oanda.com:443/v3/instruments/EUR_USD/candles?price=M&granularity=S5&from=2025-04-07T02%3A48%3A06.747334%2B00%3A00.000000000Z&to=2025-04-07T08%3A48%3A06.747334%2B00%3A00.000000000Z does not have field 'candles' (contains 'errorMessage')\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "\n",
      "Loop Iteration: 8\n",
      "Fetching history from 2025-04-07 02:48:18 UTC to 2025-04-12 02:48:18 UTC\n",
      "ERROR during get_history: 400 response for GET https://api-fxpractice.oanda.com:443/v3/instruments/EUR_USD/candles?price=M&granularity=S5&from=2025-04-07T02%3A48%3A18.872457%2B00%3A00.000000000Z&to=2025-04-07T08%3A48%3A18.872457%2B00%3A00.000000000Z does not have field 'candles' (contains 'errorMessage')\n",
      "      This often indicates invalid request parameters (e.g., dates).\n",
      "      *** PLEASE VERIFY YOUR SYSTEM CLOCK IS CORRECT! ***\n",
      "Waiting before retry...\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Traceback (most recent call last):\n",
      "  File \"C:\\Users\\audri\\AppData\\Local\\Temp\\ipykernel_26804\\3262540947.py\", line 53, in get_most_recent\n",
      "    df = self.get_history(instrument = self.instrument, start = past_aware, end = now_aware,\n",
      "         ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^\n",
      "  File \"C:\\Users\\audri\\anaconda3\\Lib\\site-packages\\tpqoa\\tpqoa.py\", line 232, in get_history\n",
      "    batch = self.retrieve_data(instrument, batch_start, batch_end,\n",
      "            ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^\n",
      "  File \"C:\\Users\\audri\\anaconda3\\Lib\\site-packages\\tpqoa\\tpqoa.py\", line 167, in retrieve_data\n",
      "    raw = raw.get('candles')\n",
      "          ^^^^^^^^^^^^^^^^^^\n",
      "  File \"C:\\Users\\audri\\anaconda3\\Lib\\site-packages\\v20\\response.py\", line 35, in get\n",
      "    raise ResponseNoField(self, field)\n",
      "v20.errors.ResponseNoField: 400 response for GET https://api-fxpractice.oanda.com:443/v3/instruments/EUR_USD/candles?price=M&granularity=S5&from=2025-04-07T02%3A48%3A18.872457%2B00%3A00.000000000Z&to=2025-04-07T08%3A48%3A18.872457%2B00%3A00.000000000Z does not have field 'candles' (contains 'errorMessage')\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "\n",
      "Loop Iteration: 9\n",
      "Fetching history from 2025-04-07 02:48:31 UTC to 2025-04-12 02:48:31 UTC\n",
      "ERROR during get_history: 400 response for GET https://api-fxpractice.oanda.com:443/v3/instruments/EUR_USD/candles?price=M&granularity=S5&from=2025-04-07T02%3A48%3A31.057916%2B00%3A00.000000000Z&to=2025-04-07T08%3A48%3A31.057916%2B00%3A00.000000000Z does not have field 'candles' (contains 'errorMessage')\n",
      "      This often indicates invalid request parameters (e.g., dates).\n",
      "      *** PLEASE VERIFY YOUR SYSTEM CLOCK IS CORRECT! ***\n",
      "Waiting before retry...\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Traceback (most recent call last):\n",
      "  File \"C:\\Users\\audri\\AppData\\Local\\Temp\\ipykernel_26804\\3262540947.py\", line 53, in get_most_recent\n",
      "    df = self.get_history(instrument = self.instrument, start = past_aware, end = now_aware,\n",
      "         ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^\n",
      "  File \"C:\\Users\\audri\\anaconda3\\Lib\\site-packages\\tpqoa\\tpqoa.py\", line 232, in get_history\n",
      "    batch = self.retrieve_data(instrument, batch_start, batch_end,\n",
      "            ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^\n",
      "  File \"C:\\Users\\audri\\anaconda3\\Lib\\site-packages\\tpqoa\\tpqoa.py\", line 167, in retrieve_data\n",
      "    raw = raw.get('candles')\n",
      "          ^^^^^^^^^^^^^^^^^^\n",
      "  File \"C:\\Users\\audri\\anaconda3\\Lib\\site-packages\\v20\\response.py\", line 35, in get\n",
      "    raise ResponseNoField(self, field)\n",
      "v20.errors.ResponseNoField: 400 response for GET https://api-fxpractice.oanda.com:443/v3/instruments/EUR_USD/candles?price=M&granularity=S5&from=2025-04-07T02%3A48%3A31.057916%2B00%3A00.000000000Z&to=2025-04-07T08%3A48%3A31.057916%2B00%3A00.000000000Z does not have field 'candles' (contains 'errorMessage')\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "\n",
      "Loop Iteration: 10\n",
      "Fetching history from 2025-04-07 02:48:43 UTC to 2025-04-12 02:48:43 UTC\n",
      "ERROR during get_history: 400 response for GET https://api-fxpractice.oanda.com:443/v3/instruments/EUR_USD/candles?price=M&granularity=S5&from=2025-04-07T02%3A48%3A43.188303%2B00%3A00.000000000Z&to=2025-04-07T08%3A48%3A43.188303%2B00%3A00.000000000Z does not have field 'candles' (contains 'errorMessage')\n",
      "      This often indicates invalid request parameters (e.g., dates).\n",
      "      *** PLEASE VERIFY YOUR SYSTEM CLOCK IS CORRECT! ***\n",
      "Waiting before retry...\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Traceback (most recent call last):\n",
      "  File \"C:\\Users\\audri\\AppData\\Local\\Temp\\ipykernel_26804\\3262540947.py\", line 53, in get_most_recent\n",
      "    df = self.get_history(instrument = self.instrument, start = past_aware, end = now_aware,\n",
      "         ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^\n",
      "  File \"C:\\Users\\audri\\anaconda3\\Lib\\site-packages\\tpqoa\\tpqoa.py\", line 232, in get_history\n",
      "    batch = self.retrieve_data(instrument, batch_start, batch_end,\n",
      "            ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^\n",
      "  File \"C:\\Users\\audri\\anaconda3\\Lib\\site-packages\\tpqoa\\tpqoa.py\", line 167, in retrieve_data\n",
      "    raw = raw.get('candles')\n",
      "          ^^^^^^^^^^^^^^^^^^\n",
      "  File \"C:\\Users\\audri\\anaconda3\\Lib\\site-packages\\v20\\response.py\", line 35, in get\n",
      "    raise ResponseNoField(self, field)\n",
      "v20.errors.ResponseNoField: 400 response for GET https://api-fxpractice.oanda.com:443/v3/instruments/EUR_USD/candles?price=M&granularity=S5&from=2025-04-07T02%3A48%3A43.188303%2B00%3A00.000000000Z&to=2025-04-07T08%3A48%3A43.188303%2B00%3A00.000000000Z does not have field 'candles' (contains 'errorMessage')\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "\n",
      "Loop Iteration: 11\n",
      "DEBUG: Exiting loop after 10 iterations. Check API errors or clock.\n",
      "Trading initialization failed: Failed to get recent data after 10 attempts. Check system clock and OANDA connection.\n"
     ]
    }
   ],
   "source": [
    "class ConTrader(tpqoa.tpqoa):\n",
    "    def __init__(self, conf_file, instrument, bar_length, window, units):\n",
    "        super().__init__(conf_file)\n",
    "        self.instrument = instrument\n",
    "        # Ensure bar_length is a Timedelta\n",
    "        if isinstance(bar_length, str):\n",
    "             self.bar_length = pd.to_timedelta(bar_length)\n",
    "        else:\n",
    "             self.bar_length = bar_length # Assume it's already a Timedelta if not string\n",
    "\n",
    "        self.tick_data = pd.DataFrame()\n",
    "        self.raw_data = None # Will store historical data (index might be aware initially)\n",
    "        self.data = None      # Will store data with strategy signals (index likely naive)\n",
    "        self.last_bar = None  # Stores the timestamp of the last full bar (NAIVE UTC)\n",
    "        self.units = units\n",
    "        self.position = 0\n",
    "        self.profits = []\n",
    "\n",
    "        #*****************add strategy-specific attributes here******************\n",
    "        self.window = window\n",
    "        #************************************************************************\n",
    "\n",
    "    def get_most_recent(self, days = 5):\n",
    "        print(\"\\n--- Entering get_most_recent ---\") # DEBUG\n",
    "        loop_counter = 0 # DEBUG counter\n",
    "        max_loops = 10 # DEBUG limit - Increase if needed, but fix clock first!\n",
    "\n",
    "        while True:\n",
    "            loop_counter += 1 # DEBUG\n",
    "            print(f\"\\nLoop Iteration: {loop_counter}\") # DEBUG\n",
    "\n",
    "            if loop_counter > max_loops: # DEBUG safety break\n",
    "                print(f\"DEBUG: Exiting loop after {max_loops} iterations. Check API errors or clock.\")\n",
    "                # Crucial: Raise an error or handle the failure state properly\n",
    "                # If this fails, the bot cannot initialize properly.\n",
    "                raise SystemExit(f\"Failed to get recent data after {max_loops} attempts. Check system clock and OANDA connection.\")\n",
    "                # break # Avoid breaking without setting data if possible\n",
    "\n",
    "            time.sleep(2) # Prevent rapid-fire requests on error\n",
    "\n",
    "            # Use aware UTC timestamps for API requests\n",
    "            # *** VERIFY YOUR SYSTEM CLOCK IS CORRECT ***\n",
    "            now_aware = datetime.now(timezone.utc)\n",
    "            past_aware = now_aware - timedelta(days = days)\n",
    "\n",
    "            # Format for printing/logging clarity\n",
    "            now_str = now_aware.strftime('%Y-%m-%d %H:%M:%S %Z')\n",
    "            past_str = past_aware.strftime('%Y-%m-%d %H:%M:%S %Z')\n",
    "            print(f\"Fetching history from {past_str} to {now_str}\") # DEBUG API call times\n",
    "\n",
    "            try:\n",
    "                # Request data - localize=False is default and usually returns UTC anyway\n",
    "                df = self.get_history(instrument = self.instrument, start = past_aware, end = now_aware,\n",
    "                                      granularity = \"S5\", price = \"M\", localize = False)\n",
    "\n",
    "                # Check if the response was successful (tpqoa might handle some errors internally,\n",
    "                # but the ResponseNoField error shows it doesn't catch everything)\n",
    "                # A more robust check might involve inspecting the response before accessing 'c'\n",
    "                if not isinstance(df, pd.DataFrame) or 'c' not in df.columns:\n",
    "                     print(f\"Warning: get_history did not return expected DataFrame format. Response: {df}\")\n",
    "                     # Consider adding more specific error handling based on tpqoa's return types on error\n",
    "                     continue\n",
    "\n",
    "                df = df.c.dropna().to_frame() # Select close price, drop NaNs\n",
    "\n",
    "                if df.empty:\n",
    "                    print(\"Warning: get_history returned DataFrame with no valid close prices. Waiting and retrying.\") # DEBUG\n",
    "                    continue # Skip rest of loop and try again\n",
    "\n",
    "                df.rename(columns = {\"c\":self.instrument}, inplace = True)\n",
    "                print(f\"  History data shape BEFORE resample: {df.shape}\")\n",
    "                if not df.empty:\n",
    "                    print(f\"  History data index TZ before resample: {df.index.tz}\") # DEBUG Check TZ (Should be UTC)\n",
    "                    # print(f\"  History data tail BEFORE resample:\\n{df.tail(3)}\") # Optional Debug\n",
    "\n",
    "                # Resample to desired bar length. label='right' aligns the timestamp to the end of the bar.\n",
    "                # The index should remain timezone-aware (UTC) after resampling.\n",
    "                # Drop the last bar? -> Reconsider this. The loop condition handles waiting.\n",
    "                # Let's keep the last bar for now unless there's a strong reason not to.\n",
    "                # df_resampled = df.resample(self.bar_length, label = \"right\").last().dropna().iloc[:-1] <--- Original\n",
    "                df_resampled = df.resample(self.bar_length, label = \"right\").last().dropna() # Keep all bars\n",
    "\n",
    "                print(f\"  Resampled data shape (after dropna): {df_resampled.shape}\")\n",
    "                if df_resampled.empty:\n",
    "                    print(\"Warning: Resampled DataFrame is empty after dropna. Waiting and retrying.\")\n",
    "                    continue\n",
    "                print(f\"  Resampled data index TZ after resample: {df_resampled.index.tz}\") # DEBUG Check TZ (Should be UTC)\n",
    "                # print(f\"  Resampled data tail:\\n{df_resampled.tail(3)}\") # Optional Debug\n",
    "\n",
    "                # Store the raw resampled data (index is aware UTC)\n",
    "                self.raw_data = df_resampled.copy()\n",
    "\n",
    "                # --- Set last_bar to NAIVE UTC ---\n",
    "                # This is a key decision: Internal logic will use naive UTC for comparisons.\n",
    "                if not self.raw_data.empty:\n",
    "                    aware_last_bar = self.raw_data.index[-1] # This is TZ-AWARE UTC\n",
    "                    self.last_bar = aware_last_bar.tz_localize(None) # Convert to NAIVE UTC\n",
    "                    print(f\"DEBUG: Initial self.last_bar set to NAIVE UTC: {self.last_bar}\")\n",
    "                else:\n",
    "                    print(\"WARNING: raw_data is empty after resample/dropna, cannot set last_bar. Retrying.\")\n",
    "                    continue\n",
    "\n",
    "                # --- Check if the last bar is recent enough ---\n",
    "                # Use consistent NAIVE UTC times for the loop's exit check\n",
    "                current_check_time_naive = datetime.now(timezone.utc).replace(tzinfo=None)\n",
    "                if self.last_bar is None: # Should not happen if code above worked\n",
    "                    print(\"WARNING: self.last_bar is None before time diff check. Retrying.\")\n",
    "                    continue\n",
    "\n",
    "                # Perform comparison with naive timestamps\n",
    "                time_diff = current_check_time_naive - self.last_bar\n",
    "\n",
    "                print(f\"  Current Naive UTC Time:    {current_check_time_naive}\")\n",
    "                print(f\"  Last Bar Time (Naive UTC): {self.last_bar}\")\n",
    "                print(f\"  Bar Length:                {self.bar_length}\")\n",
    "                print(f\"  Time Difference:           {time_diff}\")\n",
    "                print(f\"  Is Time Diff < Bar Length? {time_diff < self.bar_length}\")\n",
    "\n",
    "                # Exit loop if the last bar is recent enough (within one bar_length of current time)\n",
    "                if time_diff < self.bar_length:\n",
    "                    print(\"--- Historical data up-to-date enough. Exiting get_most_recent loop. ---\") # DEBUG\n",
    "                    break # Exit the while loop\n",
    "\n",
    "                else:\n",
    "                    print(\"--- Last bar is older than bar_length. Looping again to get more recent data. ---\")\n",
    "\n",
    "\n",
    "            except v20.errors.ResponseNoField as e:\n",
    "                 # Catch the specific error from the traceback\n",
    "                 print(f\"ERROR during get_history: {e}\")\n",
    "                 print(\"      This often indicates invalid request parameters (e.g., dates).\")\n",
    "                 print(\"      *** PLEASE VERIFY YOUR SYSTEM CLOCK IS CORRECT! ***\")\n",
    "                 traceback.print_exc() # Print full traceback\n",
    "                 print(\"Waiting before retry...\")\n",
    "                 time.sleep(10) # Wait longer after API errors\n",
    "\n",
    "            except Exception as e:\n",
    "                print(f\"UNEXPECTED ERROR during get_history or processing: {e}\") # DEBUG\n",
    "                traceback.print_exc() # Print full traceback for debugging\n",
    "                print(\"Waiting before retry...\")\n",
    "                time.sleep(5)\n",
    "\n",
    "        print(\"--- Exiting get_most_recent ---\") # DEBUG\n",
    "\n",
    "    def start_trading(self, days): # NEW\n",
    "        try:\n",
    "            self.get_most_recent(days)\n",
    "            # Only start streaming if get_most_recent was successful (didn't raise SystemExit)\n",
    "            print(\"\\n--- Starting Data Stream ---\")\n",
    "            self.stream_data(self.instrument) # This will block until stopped or error\n",
    "            print(\"\\n--- Data Stream Stopped ---\")\n",
    "        except SystemExit as e:\n",
    "             print(f\"Trading initialization failed: {e}\")\n",
    "        except Exception as e:\n",
    "             print(f\"An error occurred during trading: {e}\")\n",
    "             traceback.print_exc()\n",
    "\n",
    "\n",
    "    def on_success(self, time_str, bid, ask): # time is likely a string from stream\n",
    "        # Add a counter print inside the callback if needed\n",
    "        # print(f\"Tick: {self.ticks} | Time: {time_str} | Bid: {bid} | Ask: {ask}\", end='\\r', flush=True)\n",
    "\n",
    "        if self.stop_stream: # Check if termination was requested\n",
    "             print(\"Stop stream flag is set, skipping tick processing.\")\n",
    "             return\n",
    "\n",
    "        # --- Convert incoming time string to NAIVE UTC Timestamp ---\n",
    "        try:\n",
    "            # OANDA streams provide time in RFC3339 format (UTC)\n",
    "            # pd.to_datetime usually handles this well and makes it tz-aware (UTC)\n",
    "            aware_tick_time = pd.to_datetime(time_str)\n",
    "            # Convert to naive UTC to match self.last_bar for comparisons\n",
    "            recent_tick = aware_tick_time.tz_localize(None) # Make naive\n",
    "        except Exception as e:\n",
    "             print(f\"\\nError converting stream timestamp '{time_str}': {e}\")\n",
    "             # Decide how to handle: skip tick, use last known good time? For now, skip.\n",
    "             return\n",
    "\n",
    "        # Define stop - Check this *before* expensive operations if possible\n",
    "        # Adjust the number of ticks as needed for testing\n",
    "        if self.ticks >= 500: # Reduced for quicker testing if desired\n",
    "            print(\"\\nDEBUG: Reached tick limit.\")\n",
    "            if not self.stop_stream: # Ensure terminate_session is called only once\n",
    "                 self.terminate_session(cause = f\"Scheduled Session End after {self.ticks} ticks.\")\n",
    "            return # Exit callback\n",
    "\n",
    "        # --- Check if last_bar is initialized ---\n",
    "        # This check prevents the TypeError if get_most_recent failed silently\n",
    "        if self.last_bar is None:\n",
    "            print(\"\\nERROR: self.last_bar is None. Cannot process ticks. Historical data fetch likely failed.\")\n",
    "            # Consider stopping the stream here if this state persists\n",
    "            if not self.stop_stream:\n",
    "                self.terminate_session(cause=\"Initialization Error: last_bar not set.\")\n",
    "            return\n",
    "\n",
    "        # Collect and store tick data (using naive UTC index)\n",
    "        df = pd.DataFrame({self.instrument:(ask + bid)/2},\n",
    "                          index = [recent_tick])\n",
    "        # Use pd.concat - consider efficiency for very high frequency\n",
    "        self.tick_data = pd.concat([self.tick_data, df])\n",
    "\n",
    "        # --- Check if a new bar needs to be formed ---\n",
    "        # Compare naive recent_tick with naive self.last_bar\n",
    "        try:\n",
    "             time_since_last_bar = recent_tick - self.last_bar\n",
    "             # print(f\"Tick time: {recent_tick}, Last Bar: {self.last_bar}, Diff: {time_since_last_bar}, BarLength: {self.bar_length}\") # Fine-grained Debug\n",
    "             if time_since_last_bar >= self.bar_length:\n",
    "                 # DEBUG: Log when resampling is triggered\n",
    "                 print(f\"\\nDEBUG: Triggering resample. Recent Tick: {recent_tick}, Last Bar: {self.last_bar}, Diff: {time_since_last_bar}\")\n",
    "                 self.resample_and_join()\n",
    "                 self.define_strategy()\n",
    "                 self.execute_trades()\n",
    "        except TypeError as e:\n",
    "             # This could happen if somehow last_bar becomes None *after* the initial check\n",
    "             print(f\"\\nERROR comparing timestamps: {e}. recent_tick={recent_tick} ({type(recent_tick)}), self.last_bar={self.last_bar} ({type(self.last_bar)})\")\n",
    "             if not self.stop_stream:\n",
    "                self.terminate_session(cause=\"Timestamp comparison error during stream.\")\n",
    "             return\n",
    "        # REMOVED Duplicate check block that was here\n",
    "\n",
    "\n",
    "    def resample_and_join(self):\n",
    "        print(\"DEBUG: Entering resample_and_join\")\n",
    "        if self.tick_data.empty:\n",
    "             print(\"DEBUG: No tick data to resample.\")\n",
    "             return\n",
    "\n",
    "        # Resample tick_data (which has a naive UTC index)\n",
    "        # label='right' aligns the timestamp. last() gets the last tick in the interval.\n",
    "        # Using iloc[:-1] drops the most recent, potentially incomplete, bar formed from ticks.\n",
    "        # Using ffill() before dropping seems unusual - normally you'd resample, maybe drop, then maybe fill.\n",
    "        # Let's stick to the original logic for now, but be aware.\n",
    "        resampled_ticks = self.tick_data.resample(self.bar_length,\n",
    "                                                  label=\"right\").last().ffill().iloc[:-1] # Original logic\n",
    "        # Alternative: resampled_ticks = self.tick_data.resample(self.bar_length, label=\"right\").last().iloc[:-1] # Drop potentially incomplete last bar\n",
    "\n",
    "\n",
    "        print(f\"DEBUG: Resampled ticks shape: {resampled_ticks.shape}\")\n",
    "        print(f\"DEBUG: Resampled ticks index TZ: {resampled_ticks.index.tz}\") # Should be None (naive)\n",
    "\n",
    "        if resampled_ticks.empty:\n",
    "             print(\"DEBUG: No full bars formed from tick data resampling.\")\n",
    "             # Update last_bar from the *original* tick data if no new bars formed?\n",
    "             # Or just keep the old last_bar. Let's keep old one for now.\n",
    "             # We still need to clear the processed ticks.\n",
    "             # Find the timestamp of the *last* tick that was considered in the (empty) resample\n",
    "             # This is tricky. A simpler approach might be to just clear ticks up to the last *potential* bar end.\n",
    "             # Clear ticks that are older than the start of the *next* potential bar\n",
    "             # potential_next_bar_start = self.last_bar + self.bar_length\n",
    "             # self.tick_data = self.tick_data[self.tick_data.index >= potential_next_bar_start]\n",
    "             # Simpler for now: just clear all ticks if no new bar, although this might lose edge ticks.\n",
    "             # A safer but complex way: determine the timestamp of the last tick used by resample, clear before that.\n",
    "             # Let's reset based on the last successful bar for simplicity, expecting ticks will catch up.\n",
    "             self.tick_data = self.tick_data[self.tick_data.index > self.last_bar]\n",
    "             return # No new bars to add to raw_data\n",
    "\n",
    "        # --- Ensure raw_data index is NAIVE UTC before concatenating ---\n",
    "        # self.raw_data's index *should* be aware UTC from get_history initially.\n",
    "        if self.raw_data.index.tz is not None:\n",
    "            print(\"DEBUG: Converting raw_data index to naive UTC before concat.\")\n",
    "            self.raw_data.index = self.raw_data.index.tz_localize(None)\n",
    "        elif self.raw_data.index.tz is None:\n",
    "             # If it's already naive (e.g., after a previous resample_and_join), do nothing.\n",
    "             # print(\"DEBUG: raw_data index is already naive.\")\n",
    "             pass\n",
    "\n",
    "\n",
    "        # Concatenate historical (now naive) data with new bars from ticks (naive)\n",
    "        self.raw_data = pd.concat([self.raw_data, resampled_ticks])\n",
    "\n",
    "        # Ensure no duplicate indices if resampling overlaps - keep the newest data\n",
    "        self.raw_data = self.raw_data[~self.raw_data.index.duplicated(keep='last')]\n",
    "        print(f\"DEBUG: raw_data shape after concat: {self.raw_data.shape}\")\n",
    "\n",
    "\n",
    "        # --- Update last_bar to the latest timestamp in raw_data (must be NAIVE UTC) ---\n",
    "        if not self.raw_data.empty:\n",
    "            new_last_bar = self.raw_data.index[-1]\n",
    "            # Ensure it's naive (it should be after the concat logic above)\n",
    "            if hasattr(new_last_bar, 'tzinfo') and new_last_bar.tzinfo is not None:\n",
    "                 print(f\"WARNING: new_last_bar was aware ({new_last_bar.tzinfo}) after concat, converting.\")\n",
    "                 self.last_bar = new_last_bar.tz_localize(None) # Convert to NAIVE UTC\n",
    "            else:\n",
    "                 self.last_bar = new_last_bar # Already naive\n",
    "\n",
    "            print(f\"DEBUG: Updated self.last_bar (Naive UTC): {self.last_bar}\")\n",
    "\n",
    "            # Clean up tick_data: Keep only ticks that are strictly newer than the new last bar\n",
    "            # This prevents reprocessing ticks that formed the bar we just added.\n",
    "            self.tick_data = self.tick_data[self.tick_data.index > self.last_bar]\n",
    "            print(f\"DEBUG: Remaining tick_data size after cleanup: {self.tick_data.shape[0]}\")\n",
    "        else:\n",
    "             # This case should ideally not happen if checks above work\n",
    "             print(\"WARNING: raw_data became empty during resample_and_join.\")\n",
    "             # What should last_bar be? Maybe revert to the last tick time? Risky.\n",
    "             # Keep the old self.last_bar? Or set to None and handle? Let's keep old one.\n",
    "             print(f\"DEBUG: Keeping previous self.last_bar: {self.last_bar}\")\n",
    "             # Clear tick data based on the old bar\n",
    "             self.tick_data = self.tick_data[self.tick_data.index > self.last_bar]\n",
    "\n",
    "\n",
    "    def define_strategy(self): # \"strategy-specific\"\n",
    "        # Make a copy to avoid modifying raw_data unintentionally\n",
    "        # The index of self.raw_data should be naive UTC at this point\n",
    "        df = self.raw_data.copy()\n",
    "        print(f\"DEBUG: Defining strategy on data with index TZ: {df.index.tz}\") # Should be None\n",
    "\n",
    "        #******************** define your strategy here ************************\n",
    "        df[\"returns\"] = np.log(df[self.instrument] / df[self.instrument].shift())\n",
    "        # Ensure window is > 0 for rolling calculation\n",
    "        if self.window > 0 :\n",
    "             df[\"sma\"] = df[self.instrument].rolling(self.window).mean() # Optional: For context\n",
    "             df[\"position\"] = -np.sign(df.returns.rolling(self.window).mean())\n",
    "             # Handle potential NaN from rolling mean in the position column\n",
    "             df['position'] = df['position'].ffill().fillna(0) # Forward fill, then fill remaining NaN with 0 (neutral)\n",
    "        else:\n",
    "             # Handle cases where window might be 0 or negative (e.g., default to neutral)\n",
    "             print(\"WARN: Window is <= 0, cannot calculate rolling mean. Setting position to 0.\")\n",
    "             df[\"sma\"] = df[self.instrument] # Or NaN, or skip\n",
    "             df[\"position\"] = 0\n",
    "        #***********************************************************************\n",
    "\n",
    "        self.data = df.copy()\n",
    "        # print(f\"DEBUG: Strategy data tail:\\n{self.data.tail()}\") # Optional debug\n",
    "\n",
    "    def execute_trades(self):\n",
    "        # Ensure data and position column exist and have values\n",
    "        if self.data is None or 'position' not in self.data.columns or self.data.empty:\n",
    "             print(\"WARN: Strategy data not ready, cannot execute trades.\")\n",
    "             return\n",
    "\n",
    "        # Get the latest signal\n",
    "        signal = self.data[\"position\"].iloc[-1]\n",
    "\n",
    "        print(f\"DEBUG: Executing trades. Current Position: {self.position}, Target Signal: {signal}\")\n",
    "\n",
    "        # Target position is LONG\n",
    "        if signal == 1:\n",
    "            if self.position == 0:\n",
    "                print(\"Trade: Going Long\")\n",
    "                order = self.create_order(self.instrument, self.units, suppress = True, ret = True)\n",
    "                self.report_trade(order, \"GOING LONG\")\n",
    "                self.position = 1 # Update position based on successful order (?) - create_order doesn't confirm fill\n",
    "            elif self.position == -1:\n",
    "                print(\"Trade: Closing Short, Going Long\")\n",
    "                order = self.create_order(self.instrument, self.units * 2, suppress = True, ret = True)\n",
    "                self.report_trade(order, \"GOING LONG (closing short)\")\n",
    "                self.position = 1\n",
    "            # else: position is already 1, do nothing\n",
    "\n",
    "        # Target position is SHORT\n",
    "        elif signal == -1:\n",
    "            if self.position == 0:\n",
    "                print(\"Trade: Going Short\")\n",
    "                order = self.create_order(self.instrument, -self.units, suppress = True, ret = True)\n",
    "                self.report_trade(order, \"GOING SHORT\")\n",
    "                self.position = -1\n",
    "            elif self.position == 1:\n",
    "                print(\"Trade: Closing Long, Going Short\")\n",
    "                order = self.create_order(self.instrument, -self.units * 2, suppress = True, ret = True)\n",
    "                self.report_trade(order, \"GOING SHORT (closing long)\")\n",
    "                self.position = -1\n",
    "            # else: position is already -1, do nothing\n",
    "\n",
    "        # Target position is NEUTRAL\n",
    "        elif signal == 0:\n",
    "            if self.position == -1:\n",
    "                print(\"Trade: Closing Short, Going Neutral\")\n",
    "                order = self.create_order(self.instrument, self.units, suppress = True, ret = True)\n",
    "                self.report_trade(order, \"GOING NEUTRAL (closing short)\")\n",
    "                self.position = 0\n",
    "            elif self.position == 1:\n",
    "                print(\"Trade: Closing Long, Going Neutral\")\n",
    "                order = self.create_order(self.instrument, -self.units, suppress = True, ret = True)\n",
    "                self.report_trade(order, \"GOING NEUTRAL (closing long)\")\n",
    "                self.position = 0\n",
    "            # else: position is already 0, do nothing\n",
    "\n",
    "    def report_trade(self, order, going):\n",
    "        # Check if order is valid (tpqoa might return None or an error dict on failure)\n",
    "        if order is None or not isinstance(order, dict) or \"time\" not in order:\n",
    "             print(\"\\n\" + 100* \"-\")\n",
    "             print(f\"TRADE REPORT: Order failed or invalid response for action '{going}'. Order: {order}\")\n",
    "             print(100 * \"-\" + \"\\n\")\n",
    "             # Should not update self.position if order failed.\n",
    "             # The execute_trades logic currently assumes success. More robust logic\n",
    "             # would confirm the order fill status before updating self.position.\n",
    "             return\n",
    "\n",
    "        time = order.get(\"time\", \"N/A\") # Use .get for safety\n",
    "        units = order.get(\"units\", \"N/A\")\n",
    "        price = order.get(\"price\", \"N/A\")\n",
    "        pl = float(order.get(\"pl\", 0.0)) # Use .get and ensure float conversion\n",
    "        self.profits.append(pl)\n",
    "        cumpl = sum(self.profits)\n",
    "        print(\"\\n\" + 100* \"-\")\n",
    "        print(f\"{time} | {going}\")\n",
    "        # Use f-string for cleaner formatting\n",
    "        print(f\"{time} | units = {units} | price = {price} | P&L = {pl:.2f} | Cum P&L = {cumpl:.2f}\")\n",
    "        print(100 * \"-\" + \"\\n\")\n",
    "\n",
    "    def terminate_session(self, cause): # NEW\n",
    "        print(f\"\\nTerminating Session: {cause}\")\n",
    "        if not hasattr(self, 'stop_stream') or self.stop_stream:\n",
    "             print(\"Termination already in progress or stream not started.\")\n",
    "             return # Avoid duplicate actions\n",
    "\n",
    "        self.stop_stream = True # Signal to stop processing ticks in on_success\n",
    "\n",
    "        # Close any open position\n",
    "        if self.position != 0:\n",
    "            print(f\"Closing position: {self.position * self.units} units of {self.instrument}\")\n",
    "            close_units = -self.position * self.units\n",
    "            try:\n",
    "                # Adding a timeout might be wise for cleanup orders\n",
    "                 close_order = self.create_order(self.instrument, units=close_units,\n",
    "                                                suppress=True, ret=True) # Consider timeout=10\n",
    "                 self.report_trade(close_order, f\"CLOSING FINAL POSITION ({'Long' if self.position == 1 else 'Short'})\")\n",
    "                 self.position = 0 # Assume close order works for termination\n",
    "            except Exception as e:\n",
    "                 print(f\"ERROR closing final position: {e}\")\n",
    "                 # Log this error, manual intervention might be needed\n",
    "        else:\n",
    "            print(\"No open position to close.\")\n",
    "\n",
    "        # The stream might stop itself once stop_stream is true and the callback returns,\n",
    "        # or you might need a more explicit stop depending on how tpqoa handles blocking.\n",
    "        # If stream_data blocks indefinitely, you might need threading/asyncio\n",
    "        # or modify tpqoa's stream logic if possible.\n",
    "        # For now, setting stop_stream is the main mechanism provided by this class structure.\n",
    "        print(\"Termination request processed.\")\n",
    "\n",
    "\n",
    "# --- Main Execution ---\n",
    "if __name__ == \"__main__\":\n",
    "    # Ensure your oanda.cfg file is correct and has valid credentials/account ID\n",
    "    try:\n",
    "        trader = ConTrader(\"oanda.cfg\", \"EUR_USD\", bar_length=\"1min\", window = 1, units = 100)\n",
    "\n",
    "        # *** BEFORE RUNNING: CHECK YOUR SYSTEM CLOCK! ***\n",
    "\n",
    "        trader.start_trading(days = 5) # Start the process\n",
    "\n",
    "    except FileNotFoundError:\n",
    "        print(\"ERROR: oanda.cfg not found. Please ensure the configuration file is in the correct directory.\")\n",
    "    except Exception as e:\n",
    "        print(f\"An error occurred during initialization or execution: {e}\")\n",
    "        traceback.print_exc()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "id": "bd21db4e-5207-48a5-9c05-56cb299313df",
   "metadata": {},
   "outputs": [],
   "source": [
    "trader = ConTrader(\"oanda.cfg\", \"EUR_USD\", \"1min\", window = 1, units = 100)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "id": "9e1ee9a0-bd1a-4ee0-bfc5-8d9f43c73780",
   "metadata": {},
   "outputs": [],
   "source": [
    "# This execution code needs to replaced.\n",
    "\n",
    "# trader.get_most_recent()\n",
    "# trader.stream_data(trader.instrument, stop = 100)\n",
    "# if trader.position != 0:\n",
    "#     close_order = trader.create_order(trader.instrument, units = -trader.position * trader.units,\n",
    "#                                       suppress = True, ret = True) \n",
    "#     trader.report_trade(close_order, \"GOING NEUTRAL\")\n",
    "#     trader.position = 0"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "id": "9f26e4d1-bd5c-4fe6-b859-0cfbebfa15c5",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "\n",
      "--- Entering get_most_recent ---\n",
      "\n",
      "Loop Iteration: 1\n",
      "Fetching history from 2025-04-07 02:48:55 UTC to 2025-04-12 02:48:55 UTC\n",
      "ERROR during get_history: 400 response for GET https://api-fxpractice.oanda.com:443/v3/instruments/EUR_USD/candles?price=M&granularity=S5&from=2025-04-07T02%3A48%3A55.327450%2B00%3A00.000000000Z&to=2025-04-07T08%3A48%3A55.327450%2B00%3A00.000000000Z does not have field 'candles' (contains 'errorMessage')\n",
      "      This often indicates invalid request parameters (e.g., dates).\n",
      "      *** PLEASE VERIFY YOUR SYSTEM CLOCK IS CORRECT! ***\n",
      "Waiting before retry...\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Traceback (most recent call last):\n",
      "  File \"C:\\Users\\audri\\AppData\\Local\\Temp\\ipykernel_26804\\3262540947.py\", line 53, in get_most_recent\n",
      "    df = self.get_history(instrument = self.instrument, start = past_aware, end = now_aware,\n",
      "         ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^\n",
      "  File \"C:\\Users\\audri\\anaconda3\\Lib\\site-packages\\tpqoa\\tpqoa.py\", line 232, in get_history\n",
      "    batch = self.retrieve_data(instrument, batch_start, batch_end,\n",
      "            ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^\n",
      "  File \"C:\\Users\\audri\\anaconda3\\Lib\\site-packages\\tpqoa\\tpqoa.py\", line 167, in retrieve_data\n",
      "    raw = raw.get('candles')\n",
      "          ^^^^^^^^^^^^^^^^^^\n",
      "  File \"C:\\Users\\audri\\anaconda3\\Lib\\site-packages\\v20\\response.py\", line 35, in get\n",
      "    raise ResponseNoField(self, field)\n",
      "v20.errors.ResponseNoField: 400 response for GET https://api-fxpractice.oanda.com:443/v3/instruments/EUR_USD/candles?price=M&granularity=S5&from=2025-04-07T02%3A48%3A55.327450%2B00%3A00.000000000Z&to=2025-04-07T08%3A48%3A55.327450%2B00%3A00.000000000Z does not have field 'candles' (contains 'errorMessage')\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "\n",
      "Loop Iteration: 2\n",
      "Fetching history from 2025-04-07 02:49:07 UTC to 2025-04-12 02:49:07 UTC\n",
      "ERROR during get_history: 400 response for GET https://api-fxpractice.oanda.com:443/v3/instruments/EUR_USD/candles?price=M&granularity=S5&from=2025-04-07T02%3A49%3A07.711207%2B00%3A00.000000000Z&to=2025-04-07T08%3A49%3A07.711207%2B00%3A00.000000000Z does not have field 'candles' (contains 'errorMessage')\n",
      "      This often indicates invalid request parameters (e.g., dates).\n",
      "      *** PLEASE VERIFY YOUR SYSTEM CLOCK IS CORRECT! ***\n",
      "Waiting before retry...\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Traceback (most recent call last):\n",
      "  File \"C:\\Users\\audri\\AppData\\Local\\Temp\\ipykernel_26804\\3262540947.py\", line 53, in get_most_recent\n",
      "    df = self.get_history(instrument = self.instrument, start = past_aware, end = now_aware,\n",
      "         ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^\n",
      "  File \"C:\\Users\\audri\\anaconda3\\Lib\\site-packages\\tpqoa\\tpqoa.py\", line 232, in get_history\n",
      "    batch = self.retrieve_data(instrument, batch_start, batch_end,\n",
      "            ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^\n",
      "  File \"C:\\Users\\audri\\anaconda3\\Lib\\site-packages\\tpqoa\\tpqoa.py\", line 167, in retrieve_data\n",
      "    raw = raw.get('candles')\n",
      "          ^^^^^^^^^^^^^^^^^^\n",
      "  File \"C:\\Users\\audri\\anaconda3\\Lib\\site-packages\\v20\\response.py\", line 35, in get\n",
      "    raise ResponseNoField(self, field)\n",
      "v20.errors.ResponseNoField: 400 response for GET https://api-fxpractice.oanda.com:443/v3/instruments/EUR_USD/candles?price=M&granularity=S5&from=2025-04-07T02%3A49%3A07.711207%2B00%3A00.000000000Z&to=2025-04-07T08%3A49%3A07.711207%2B00%3A00.000000000Z does not have field 'candles' (contains 'errorMessage')\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "\n",
      "Loop Iteration: 3\n",
      "Fetching history from 2025-04-07 02:49:19 UTC to 2025-04-12 02:49:19 UTC\n",
      "ERROR during get_history: 400 response for GET https://api-fxpractice.oanda.com:443/v3/instruments/EUR_USD/candles?price=M&granularity=S5&from=2025-04-07T02%3A49%3A19.840794%2B00%3A00.000000000Z&to=2025-04-07T08%3A49%3A19.840794%2B00%3A00.000000000Z does not have field 'candles' (contains 'errorMessage')\n",
      "      This often indicates invalid request parameters (e.g., dates).\n",
      "      *** PLEASE VERIFY YOUR SYSTEM CLOCK IS CORRECT! ***\n",
      "Waiting before retry...\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Traceback (most recent call last):\n",
      "  File \"C:\\Users\\audri\\AppData\\Local\\Temp\\ipykernel_26804\\3262540947.py\", line 53, in get_most_recent\n",
      "    df = self.get_history(instrument = self.instrument, start = past_aware, end = now_aware,\n",
      "         ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^\n",
      "  File \"C:\\Users\\audri\\anaconda3\\Lib\\site-packages\\tpqoa\\tpqoa.py\", line 232, in get_history\n",
      "    batch = self.retrieve_data(instrument, batch_start, batch_end,\n",
      "            ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^\n",
      "  File \"C:\\Users\\audri\\anaconda3\\Lib\\site-packages\\tpqoa\\tpqoa.py\", line 167, in retrieve_data\n",
      "    raw = raw.get('candles')\n",
      "          ^^^^^^^^^^^^^^^^^^\n",
      "  File \"C:\\Users\\audri\\anaconda3\\Lib\\site-packages\\v20\\response.py\", line 35, in get\n",
      "    raise ResponseNoField(self, field)\n",
      "v20.errors.ResponseNoField: 400 response for GET https://api-fxpractice.oanda.com:443/v3/instruments/EUR_USD/candles?price=M&granularity=S5&from=2025-04-07T02%3A49%3A19.840794%2B00%3A00.000000000Z&to=2025-04-07T08%3A49%3A19.840794%2B00%3A00.000000000Z does not have field 'candles' (contains 'errorMessage')\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "\n",
      "Loop Iteration: 4\n",
      "Fetching history from 2025-04-07 02:49:31 UTC to 2025-04-12 02:49:31 UTC\n",
      "ERROR during get_history: 400 response for GET https://api-fxpractice.oanda.com:443/v3/instruments/EUR_USD/candles?price=M&granularity=S5&from=2025-04-07T02%3A49%3A31.948749%2B00%3A00.000000000Z&to=2025-04-07T08%3A49%3A31.948749%2B00%3A00.000000000Z does not have field 'candles' (contains 'errorMessage')\n",
      "      This often indicates invalid request parameters (e.g., dates).\n",
      "      *** PLEASE VERIFY YOUR SYSTEM CLOCK IS CORRECT! ***\n",
      "Waiting before retry...\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Traceback (most recent call last):\n",
      "  File \"C:\\Users\\audri\\AppData\\Local\\Temp\\ipykernel_26804\\3262540947.py\", line 53, in get_most_recent\n",
      "    df = self.get_history(instrument = self.instrument, start = past_aware, end = now_aware,\n",
      "         ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^\n",
      "  File \"C:\\Users\\audri\\anaconda3\\Lib\\site-packages\\tpqoa\\tpqoa.py\", line 232, in get_history\n",
      "    batch = self.retrieve_data(instrument, batch_start, batch_end,\n",
      "            ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^\n",
      "  File \"C:\\Users\\audri\\anaconda3\\Lib\\site-packages\\tpqoa\\tpqoa.py\", line 167, in retrieve_data\n",
      "    raw = raw.get('candles')\n",
      "          ^^^^^^^^^^^^^^^^^^\n",
      "  File \"C:\\Users\\audri\\anaconda3\\Lib\\site-packages\\v20\\response.py\", line 35, in get\n",
      "    raise ResponseNoField(self, field)\n",
      "v20.errors.ResponseNoField: 400 response for GET https://api-fxpractice.oanda.com:443/v3/instruments/EUR_USD/candles?price=M&granularity=S5&from=2025-04-07T02%3A49%3A31.948749%2B00%3A00.000000000Z&to=2025-04-07T08%3A49%3A31.948749%2B00%3A00.000000000Z does not have field 'candles' (contains 'errorMessage')\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "\n",
      "Loop Iteration: 5\n",
      "Fetching history from 2025-04-07 02:49:44 UTC to 2025-04-12 02:49:44 UTC\n",
      "ERROR during get_history: 400 response for GET https://api-fxpractice.oanda.com:443/v3/instruments/EUR_USD/candles?price=M&granularity=S5&from=2025-04-07T02%3A49%3A44.071726%2B00%3A00.000000000Z&to=2025-04-07T08%3A49%3A44.071726%2B00%3A00.000000000Z does not have field 'candles' (contains 'errorMessage')\n",
      "      This often indicates invalid request parameters (e.g., dates).\n",
      "      *** PLEASE VERIFY YOUR SYSTEM CLOCK IS CORRECT! ***\n",
      "Waiting before retry...\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Traceback (most recent call last):\n",
      "  File \"C:\\Users\\audri\\AppData\\Local\\Temp\\ipykernel_26804\\3262540947.py\", line 53, in get_most_recent\n",
      "    df = self.get_history(instrument = self.instrument, start = past_aware, end = now_aware,\n",
      "         ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^\n",
      "  File \"C:\\Users\\audri\\anaconda3\\Lib\\site-packages\\tpqoa\\tpqoa.py\", line 232, in get_history\n",
      "    batch = self.retrieve_data(instrument, batch_start, batch_end,\n",
      "            ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^\n",
      "  File \"C:\\Users\\audri\\anaconda3\\Lib\\site-packages\\tpqoa\\tpqoa.py\", line 167, in retrieve_data\n",
      "    raw = raw.get('candles')\n",
      "          ^^^^^^^^^^^^^^^^^^\n",
      "  File \"C:\\Users\\audri\\anaconda3\\Lib\\site-packages\\v20\\response.py\", line 35, in get\n",
      "    raise ResponseNoField(self, field)\n",
      "v20.errors.ResponseNoField: 400 response for GET https://api-fxpractice.oanda.com:443/v3/instruments/EUR_USD/candles?price=M&granularity=S5&from=2025-04-07T02%3A49%3A44.071726%2B00%3A00.000000000Z&to=2025-04-07T08%3A49%3A44.071726%2B00%3A00.000000000Z does not have field 'candles' (contains 'errorMessage')\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "\n",
      "Loop Iteration: 6\n",
      "Fetching history from 2025-04-07 02:49:56 UTC to 2025-04-12 02:49:56 UTC\n",
      "ERROR during get_history: 400 response for GET https://api-fxpractice.oanda.com:443/v3/instruments/EUR_USD/candles?price=M&granularity=S5&from=2025-04-07T02%3A49%3A56.194953%2B00%3A00.000000000Z&to=2025-04-07T08%3A49%3A56.194953%2B00%3A00.000000000Z does not have field 'candles' (contains 'errorMessage')\n",
      "      This often indicates invalid request parameters (e.g., dates).\n",
      "      *** PLEASE VERIFY YOUR SYSTEM CLOCK IS CORRECT! ***\n",
      "Waiting before retry...\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Traceback (most recent call last):\n",
      "  File \"C:\\Users\\audri\\AppData\\Local\\Temp\\ipykernel_26804\\3262540947.py\", line 53, in get_most_recent\n",
      "    df = self.get_history(instrument = self.instrument, start = past_aware, end = now_aware,\n",
      "         ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^\n",
      "  File \"C:\\Users\\audri\\anaconda3\\Lib\\site-packages\\tpqoa\\tpqoa.py\", line 232, in get_history\n",
      "    batch = self.retrieve_data(instrument, batch_start, batch_end,\n",
      "            ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^\n",
      "  File \"C:\\Users\\audri\\anaconda3\\Lib\\site-packages\\tpqoa\\tpqoa.py\", line 167, in retrieve_data\n",
      "    raw = raw.get('candles')\n",
      "          ^^^^^^^^^^^^^^^^^^\n",
      "  File \"C:\\Users\\audri\\anaconda3\\Lib\\site-packages\\v20\\response.py\", line 35, in get\n",
      "    raise ResponseNoField(self, field)\n",
      "v20.errors.ResponseNoField: 400 response for GET https://api-fxpractice.oanda.com:443/v3/instruments/EUR_USD/candles?price=M&granularity=S5&from=2025-04-07T02%3A49%3A56.194953%2B00%3A00.000000000Z&to=2025-04-07T08%3A49%3A56.194953%2B00%3A00.000000000Z does not have field 'candles' (contains 'errorMessage')\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "\n",
      "Loop Iteration: 7\n",
      "Fetching history from 2025-04-07 02:50:08 UTC to 2025-04-12 02:50:08 UTC\n",
      "ERROR during get_history: 400 response for GET https://api-fxpractice.oanda.com:443/v3/instruments/EUR_USD/candles?price=M&granularity=S5&from=2025-04-07T02%3A50%3A08.370899%2B00%3A00.000000000Z&to=2025-04-07T08%3A50%3A08.370899%2B00%3A00.000000000Z does not have field 'candles' (contains 'errorMessage')\n",
      "      This often indicates invalid request parameters (e.g., dates).\n",
      "      *** PLEASE VERIFY YOUR SYSTEM CLOCK IS CORRECT! ***\n",
      "Waiting before retry...\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Traceback (most recent call last):\n",
      "  File \"C:\\Users\\audri\\AppData\\Local\\Temp\\ipykernel_26804\\3262540947.py\", line 53, in get_most_recent\n",
      "    df = self.get_history(instrument = self.instrument, start = past_aware, end = now_aware,\n",
      "         ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^\n",
      "  File \"C:\\Users\\audri\\anaconda3\\Lib\\site-packages\\tpqoa\\tpqoa.py\", line 232, in get_history\n",
      "    batch = self.retrieve_data(instrument, batch_start, batch_end,\n",
      "            ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^\n",
      "  File \"C:\\Users\\audri\\anaconda3\\Lib\\site-packages\\tpqoa\\tpqoa.py\", line 167, in retrieve_data\n",
      "    raw = raw.get('candles')\n",
      "          ^^^^^^^^^^^^^^^^^^\n",
      "  File \"C:\\Users\\audri\\anaconda3\\Lib\\site-packages\\v20\\response.py\", line 35, in get\n",
      "    raise ResponseNoField(self, field)\n",
      "v20.errors.ResponseNoField: 400 response for GET https://api-fxpractice.oanda.com:443/v3/instruments/EUR_USD/candles?price=M&granularity=S5&from=2025-04-07T02%3A50%3A08.370899%2B00%3A00.000000000Z&to=2025-04-07T08%3A50%3A08.370899%2B00%3A00.000000000Z does not have field 'candles' (contains 'errorMessage')\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "\n",
      "Loop Iteration: 8\n",
      "Fetching history from 2025-04-07 02:50:20 UTC to 2025-04-12 02:50:20 UTC\n",
      "ERROR during get_history: 400 response for GET https://api-fxpractice.oanda.com:443/v3/instruments/EUR_USD/candles?price=M&granularity=S5&from=2025-04-07T02%3A50%3A20.503094%2B00%3A00.000000000Z&to=2025-04-07T08%3A50%3A20.503094%2B00%3A00.000000000Z does not have field 'candles' (contains 'errorMessage')\n",
      "      This often indicates invalid request parameters (e.g., dates).\n",
      "      *** PLEASE VERIFY YOUR SYSTEM CLOCK IS CORRECT! ***\n",
      "Waiting before retry...\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Traceback (most recent call last):\n",
      "  File \"C:\\Users\\audri\\AppData\\Local\\Temp\\ipykernel_26804\\3262540947.py\", line 53, in get_most_recent\n",
      "    df = self.get_history(instrument = self.instrument, start = past_aware, end = now_aware,\n",
      "         ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^\n",
      "  File \"C:\\Users\\audri\\anaconda3\\Lib\\site-packages\\tpqoa\\tpqoa.py\", line 232, in get_history\n",
      "    batch = self.retrieve_data(instrument, batch_start, batch_end,\n",
      "            ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^\n",
      "  File \"C:\\Users\\audri\\anaconda3\\Lib\\site-packages\\tpqoa\\tpqoa.py\", line 167, in retrieve_data\n",
      "    raw = raw.get('candles')\n",
      "          ^^^^^^^^^^^^^^^^^^\n",
      "  File \"C:\\Users\\audri\\anaconda3\\Lib\\site-packages\\v20\\response.py\", line 35, in get\n",
      "    raise ResponseNoField(self, field)\n",
      "v20.errors.ResponseNoField: 400 response for GET https://api-fxpractice.oanda.com:443/v3/instruments/EUR_USD/candles?price=M&granularity=S5&from=2025-04-07T02%3A50%3A20.503094%2B00%3A00.000000000Z&to=2025-04-07T08%3A50%3A20.503094%2B00%3A00.000000000Z does not have field 'candles' (contains 'errorMessage')\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "\n",
      "Loop Iteration: 9\n",
      "Fetching history from 2025-04-07 02:50:32 UTC to 2025-04-12 02:50:32 UTC\n",
      "ERROR during get_history: 400 response for GET https://api-fxpractice.oanda.com:443/v3/instruments/EUR_USD/candles?price=M&granularity=S5&from=2025-04-07T02%3A50%3A32.621101%2B00%3A00.000000000Z&to=2025-04-07T08%3A50%3A32.621101%2B00%3A00.000000000Z does not have field 'candles' (contains 'errorMessage')\n",
      "      This often indicates invalid request parameters (e.g., dates).\n",
      "      *** PLEASE VERIFY YOUR SYSTEM CLOCK IS CORRECT! ***\n",
      "Waiting before retry...\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Traceback (most recent call last):\n",
      "  File \"C:\\Users\\audri\\AppData\\Local\\Temp\\ipykernel_26804\\3262540947.py\", line 53, in get_most_recent\n",
      "    df = self.get_history(instrument = self.instrument, start = past_aware, end = now_aware,\n",
      "         ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^\n",
      "  File \"C:\\Users\\audri\\anaconda3\\Lib\\site-packages\\tpqoa\\tpqoa.py\", line 232, in get_history\n",
      "    batch = self.retrieve_data(instrument, batch_start, batch_end,\n",
      "            ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^\n",
      "  File \"C:\\Users\\audri\\anaconda3\\Lib\\site-packages\\tpqoa\\tpqoa.py\", line 167, in retrieve_data\n",
      "    raw = raw.get('candles')\n",
      "          ^^^^^^^^^^^^^^^^^^\n",
      "  File \"C:\\Users\\audri\\anaconda3\\Lib\\site-packages\\v20\\response.py\", line 35, in get\n",
      "    raise ResponseNoField(self, field)\n",
      "v20.errors.ResponseNoField: 400 response for GET https://api-fxpractice.oanda.com:443/v3/instruments/EUR_USD/candles?price=M&granularity=S5&from=2025-04-07T02%3A50%3A32.621101%2B00%3A00.000000000Z&to=2025-04-07T08%3A50%3A32.621101%2B00%3A00.000000000Z does not have field 'candles' (contains 'errorMessage')\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "\n",
      "Loop Iteration: 10\n",
      "Fetching history from 2025-04-07 02:50:44 UTC to 2025-04-12 02:50:44 UTC\n",
      "ERROR during get_history: 400 response for GET https://api-fxpractice.oanda.com:443/v3/instruments/EUR_USD/candles?price=M&granularity=S5&from=2025-04-07T02%3A50%3A44.739764%2B00%3A00.000000000Z&to=2025-04-07T08%3A50%3A44.739764%2B00%3A00.000000000Z does not have field 'candles' (contains 'errorMessage')\n",
      "      This often indicates invalid request parameters (e.g., dates).\n",
      "      *** PLEASE VERIFY YOUR SYSTEM CLOCK IS CORRECT! ***\n",
      "Waiting before retry...\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Traceback (most recent call last):\n",
      "  File \"C:\\Users\\audri\\AppData\\Local\\Temp\\ipykernel_26804\\3262540947.py\", line 53, in get_most_recent\n",
      "    df = self.get_history(instrument = self.instrument, start = past_aware, end = now_aware,\n",
      "         ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^\n",
      "  File \"C:\\Users\\audri\\anaconda3\\Lib\\site-packages\\tpqoa\\tpqoa.py\", line 232, in get_history\n",
      "    batch = self.retrieve_data(instrument, batch_start, batch_end,\n",
      "            ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^\n",
      "  File \"C:\\Users\\audri\\anaconda3\\Lib\\site-packages\\tpqoa\\tpqoa.py\", line 167, in retrieve_data\n",
      "    raw = raw.get('candles')\n",
      "          ^^^^^^^^^^^^^^^^^^\n",
      "  File \"C:\\Users\\audri\\anaconda3\\Lib\\site-packages\\v20\\response.py\", line 35, in get\n",
      "    raise ResponseNoField(self, field)\n",
      "v20.errors.ResponseNoField: 400 response for GET https://api-fxpractice.oanda.com:443/v3/instruments/EUR_USD/candles?price=M&granularity=S5&from=2025-04-07T02%3A50%3A44.739764%2B00%3A00.000000000Z&to=2025-04-07T08%3A50%3A44.739764%2B00%3A00.000000000Z does not have field 'candles' (contains 'errorMessage')\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "\n",
      "Loop Iteration: 11\n",
      "DEBUG: Exiting loop after 10 iterations. Check API errors or clock.\n",
      "Trading initialization failed: Failed to get recent data after 10 attempts. Check system clock and OANDA connection.\n"
     ]
    }
   ],
   "source": [
    "trader.start_trading(days = 5)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "bcf5d0ce-1bbf-43a3-8e72-3a6ed82026f1",
   "metadata": {},
   "source": [
    "### Added Wrapper methodes for start and terminate Trading Sessions"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "0e7a450f-c804-4454-8c12-faff932688cf",
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3 (ipykernel)",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.12.4"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
